File: src/zefr_interface.cpp
#include <csignal>

#include "zefr_interface.hpp"

Zefr *ZEFR = NULL;

namespace zefr {

#ifdef _MPI
void initialize(MPI_Comm comm_in, const char *inputFile, int nGrids, int gridID, MPI_Comm world_comm)
{
  if (!ZEFR) ZEFR = new Zefr(comm_in, nGrids, gridID, world_comm);

  ZEFR->read_input(inputFile);

  auto input = ZEFR->get_input();

  if (input.catch_signals)
  {
    signal(SIGINT, signal_handler);
    signal(SIGTERM, signal_handler);
  }
}
#else
void initialize(char *input_file)
{
  if (!ZEFR) ZEFR = new Zefr();

  ZEFR->read_input(input_file);
}
#endif

void set_zefr_object(Zefr *_ZEFR)
{
  delete ZEFR;

  ZEFR = _ZEFR;
}

Zefr* get_zefr_object(void)
{
  return ZEFR;
}

void finalize(void)
{
  ZEFR->write_wall_time();
  delete ZEFR;
}

/* ---- Data-Acess Functions ---- */

BasicGeo get_basic_geo_data(void)
{
  BasicGeo geo;

  ZEFR->get_basic_geo_data(geo.btag,geo.gridType,geo.nnodes,geo.xyz,geo.iblank,geo.nwall,
                           geo.nover,geo.wallNodes,geo.overNodes,geo.nCellTypes,
                           geo.nvert_cell,geo.nface_cell,geo.nCells_type,geo.c2v);

  geo.nCellsTot = 0;
  for (int i = 0; i < geo.nCellTypes; i++)
    geo.nCellsTot += geo.nCells_type[i];

  return geo;
}

double& get_q_spt(int ele, int spt, int var)
{
  return ZEFR->get_u_spt(ele,spt,var);
}

double get_grad_spt(int ele, int spt, int dim, int var)
{
  return ZEFR->get_grad_spt(ele,spt,dim,var);
}

double& get_q_fpt(int face, int fpt, int var)
{
  return ZEFR->get_u_fpt(face,fpt,var);
}

double& get_grad_fpt(int face, int fpt, int dim, int var)
{
  return ZEFR->get_grad_fpt(face,fpt,dim,var);
}

double* get_q_spts(int &ele_stride, int &spt_stride, int &var_stride, int etype)
{
  return ZEFR->get_u_spts(ele_stride, spt_stride, var_stride, etype);
}

double* get_dq_spts(int &ele_stride, int &spt_stride, int &var_stride, int &dim_stride, int etype)
{
  return ZEFR->get_du_spts(ele_stride, spt_stride, var_stride, dim_stride, etype);
}

double* get_q_spts_d(int &ele_stride, int &spt_stride, int &var_stride, int etype)
{
  return ZEFR->get_u_spts_d(ele_stride, spt_stride, var_stride, etype);
}

double* get_dq_spts_d(int &ele_stride, int &spt_stride, int &var_stride, int &dim_stride, int etype)
{
  return ZEFR->get_du_spts_d(ele_stride, spt_stride, var_stride, dim_stride, etype);
}

ExtraGeo get_extra_geo_data(void)
{
  ExtraGeo geo;

  ZEFR->get_extra_geo_data(geo.nFaceTypes,geo.faceTypes,geo.cellTypes,
      geo.nvert_face,geo.nFaces_type,geo.f2v,geo.f2c,geo.c2f,geo.iblank_face,
      geo.iblank_cell,geo.nOverFaces,geo.overFaces,geo.nWallFaces,geo.wallFaces,
      geo.nMpiFaces,geo.mpiFaces,geo.procR,geo.mpiFidR,geo.grid_vel,geo.offset,geo.Rmat);

  geo.nFacesTot = 0;
  for (int i = 0; i < geo.nFaceTypes; i++)
    geo.nFacesTot += geo.nFaces_type[i];

  return geo;
}

GpuGeo get_gpu_geo_data(void)
{
  GpuGeo geo;

  ZEFR->get_gpu_geo_data(geo.coord_nodes,geo.coord_eles,geo.iblank_cell,geo.iblank_face);

  return geo;
}

CallbackFuncs get_callback_funcs(void)
{
  CallbackFuncs call;

  call.get_nodes_per_cell = get_nodes_per_cell;
  call.get_nodes_per_face = get_nodes_per_face;
  call.get_receptor_nodes = get_receptor_nodes;
  call.get_face_nodes = get_face_nodes;
  call.donor_inclusion_test = donor_inclusion_test;
  call.donor_frac = donor_frac;
  call.convert_to_modal = convert_to_modal;
  call.get_q_spt = get_q_spt;
  call.get_grad_spt = get_grad_spt;
  call.get_q_fpt = get_q_fpt;
  call.get_grad_fpt = get_grad_fpt;
  call.get_q_spts = get_q_spts;
  call.get_dq_spts = get_dq_spts;

  /* GPU-specific functions */
  call.fringe_data_to_device = fringe_data_to_device;
  call.unblank_data_to_device = unblank_data_to_device;
  call.get_q_spts_d = get_q_spts_d;
  call.get_dq_spts_d = get_dq_spts_d;

  call.get_face_nodes_gpu = get_face_nodes_gpu;
  call.get_cell_nodes_gpu = get_cell_nodes_gpu;

  call.get_n_weights = get_n_weights;
  call.donor_frac_gpu = donor_frac_gpu;

  return call;
}

/* ---- TIOGA Callback Functions ---- */

void get_nodes_per_cell(int* cellID, int* nNodes)
{
  ZEFR->get_nodes_per_cell(*cellID, *nNodes);
}

void get_nodes_per_face(int* faceID, int* nNodes)
{
  ZEFR->get_nodes_per_face(*faceID, *nNodes);
}

void get_receptor_nodes(int* cellID, int* nNodes, double* xyz)
{
  ZEFR->get_receptor_nodes(*cellID, *nNodes, xyz);
}

void get_face_nodes(int* faceID, int* nNodes, double* xyz)
{
  ZEFR->get_face_nodes(*faceID, *nNodes, xyz);
}

void get_face_nodes_gpu(int* faceIDs, int nFringe, int* nPtsFace, double* xyz)
{
  ZEFR->get_face_nodes_gpu(faceIDs, nFringe, nPtsFace, xyz);
}

void get_cell_nodes_gpu(int* cellIDs, int nCells, int* nPtsCell, double* xyz)
{
  ZEFR->get_cell_nodes_gpu(cellIDs, nCells, nPtsCell, xyz);
}

void donor_inclusion_test(int* cellID, double* xyz, int* passFlag, double* rst)
{
  ZEFR->donor_inclusion_test(*cellID, xyz, *passFlag, rst);
}

void donor_frac(int* cellID, double* xyz, int* nweights, int* inode,
                double* weights, double* rst, int* buffsize)
{
  ZEFR->donor_frac(*cellID, *nweights, inode, weights, rst, *buffsize);
}

int get_n_weights(int cellID)
{
  return ZEFR->get_n_weights(cellID);
}

void donor_frac_gpu(int* cellIDs, int nFringe, double* rst, double* weights)
{
  ZEFR->donor_frac_gpu(cellIDs, nFringe, rst, weights);
}

void convert_to_modal(int* cellID, int* nSpts, double* q_in, int* npts, int* index_out, double* q_out)
{
  //assert(*nSpts == *npts);
  *index_out = (*cellID) * (*nSpts);
  for (int spt = 0; spt < (*nSpts); spt++)
    q_out[spt] = q_in[spt];
}

void fringe_data_to_device(int *fringeIDs, int nFringe, int gradFlag, double *data)
{
#ifdef _GPU
  ZEFR->fringe_data_to_device(fringeIDs, nFringe, gradFlag, data);
#endif
}

void unblank_data_to_device(int *fringeIDs, int nFringe, int gradFlag, double *data)
{
#ifdef _GPU
  ZEFR->unblank_data_to_device(fringeIDs, nFringe, gradFlag, data);
#endif
}

void signal_handler(int signum)
{
  if (signum == SIGINT)
    std::cout << "Received signal SIGINT - dumping solution to file" << std::endl;
  else if (signum == SIGTERM)
    std::cout << "Received signal SIGTERM - dumping solution to file and quitting" << std::endl;
  else
    exit(signum);

  ZEFR->write_solution();

  if (signum == SIGTERM)
    exit(signum);
}

} /* namespace zefr */

File: bin/zefrInterface.py
#! /usr/bin/env python
#
# File: zefrInterface.py
# Authors: Jacob Crabill
# Last Modified: 9/02/2017


__version__ = "1.00"

# ==================================================================
# Standard Python modules

import sys
import os
import copy
import string
import types

#Extension modules
#sys.path.append(os.path.abspath(os.getenv('PYTHONPATH')))
#sys.path.append(os.path.abspath(os.getenv('PYTHONPATH')+'/numpy'))

import numpy as np

# Try to import the MPI.COMM_WORLD.module
try:
    from mpi4py import MPI
    _parallel = True

except ImportError:
    _parallel = False


try:
    import zefr
except ImportError:
    print("Import Error: ZEFR solver module")

class zefrSolver:

    def __init__(self,startfile,gridID,nGrids):
        if not os.path.isfile(startfile):
          print('Error: ZEFR input file',startfile,'not found in path')
          raise

        self.nGrids = nGrids
        self.ID = gridID
        self.inpfile=startfile

        Comm = MPI.COMM_WORLD
        rank = Comm.Get_rank()
        nproc = Comm.Get_size()

        self.gridComm = Comm.Split(gridID,rank)
        self.gridRank = gridComm.Get_rank()
        self.gridSize = gridComm.Get_size()

        # Create the directory where all ZEFR IO will occur
        if self.gridRank == 0 and os.path.isdir('zefr') == False:
          os.mkdir('zefr')

        zefr.initialize(self.gridComm,startfile,nGrids,gridID)

        # scaling for solver variable [i.e. non-dimensionalization]
        self.scale = np.array([1.,1.,1.,1.,1.],'d')

        self.z = zefr.get_zefr_object()
        self.inp = self.z.get_input()
        self.simdata = self.z.get_data()
        self.name = 'zefr'

        os.chdir('..')

    # Run one full time step
    def runStep(self,iter,nstages):
        os.chdir('zefr')

        #self.z.do_step(nstages)
        for i in range(0,nstages):
            runSubSteps(iter,i,nstages)

        os.chdir('..')

    # Run one RK stage
    def runSubSteps(self,iter,stage):
        os.chdir('zefr')

        self.z.do_rk_stage(iter,stage)

        os.chdir('..')

    def sifInitialize(self,properties,conditions):
        os.chdir('zefr') 

        rinf = properties['rinf']
        ainf = properties['ainf']

        gridScale = conditions['meshRefLength']
        Reref = conditions['reyNumber']
        Lref = conditions['reyRefLength']
        Mref = conditions['refMach']

        self.inp.rho_fs = rinf
        self.inp.L_fs = conditions['reyRefLength']

        # All conditions come in as SI units - take care of non-dim here
        dt = ainf * conditions['dt'] / gridScale
        mach_fs = conditions['Mach']
        self.restart = 'no'
        if conditions['from_restart'] == 'yes':
            self.inp.restart = 'true'
            self.inp.restartIter = conditions['restartstep']

        Re = (Reref/Lref) * gridScale / Mref * (mach_fs)
        self.forceDim = 0.5 * rinf * (refMach*ainf*gridScale)**2
        self.momDim = self.forceDim * gridScale

        # Have ZEFR do any additional setup on input parameters
        zefr.sifinit(dt,Re,mach_fs)
            
        self.scale = np.array([1./rinf,
                               1./(rinf*ainf),
                               1./(rinf*ainf),
                               1./(rinf*ainf),
                               1./(rinf*ainf*ainf),1.],'d')

        os.chdir('..')

    def getForcesAndMoments(self):
        self.z.get_forces()
        bodyForce = np.array(dptrToArray(self.simdata.forces))
        bodyForce[:3] = bodyForce[:3]*self.forceDim
        bodyForce[3:] = bodyForce[3:]*self.momDim
        return bodyForce
                        
    # sifInit called first to setup solver input
    def initData(self):
        os.chdir('zefr')

        # TODO: modify ZEFR so callbacks aren't needed yet? 
        # TODO: Or should I get callbacks here or in sifInit?

        # Setup the ZEFR solver 
        self.z.setup_solver()

        # Get all relevant geometry and solver data
        geo = zefr.get_basic_geo_data()   # Basic geometry/connectivity data
        geoAB = zefr.get_extra_geo_data() # Geo data for AB method
        cbs = zefr.get_callback_funcs()   # Callback functions for high-order/AB method
        simdata = self.simdata

        # -----------------------------------------------------------------------
        # Turn all pointers from ZEFR into (single-element) lists of numpy arrays
        # (For 2D/3D arrays, can call reshape() after the fact)
        # -----------------------------------------------------------------------
        ndims = 3
        nfields = simdata.nfields
        nspts = simdata.nspts
        ncells = geo.nCells_type
        nfaces = geoAB.nFaces_type
        nnodes = geo.nnodes
        nvert = geo.nvert_cell
        nvertf = getAB.nvert_face
        nfaces = geoAB.nFaces_type

        # Wrap all geometry data
        xyz = [zefr.dptrToArray(geo.xyz, nnodes*ndims)]
        c2v = [zefr.iptrToArray(geo.c2v, ncells*nvert)]
        wallNodes = [zefr.iptrToArray(geo.wallNodes, geo.nwall)]
        overNodes = [zefr.iptrToArray(geo.overNodes, geo.nover)]

        iblank = [zefr.iptrToArray(geo.iblank, nnodes)]
        iblank_cell = [zefr.iptrToArray(geoAB.iblank_cell, ncells)]
        iblank_face = [zefr.iptrToArray(geoAB.iblank_face, nfaces)]
        f2v = [zefr.iptrToArray(geoAB.f2v, nfaces*nvertf)]
        c2f = [zefr.iptrToArray(geoAB.c2f, ncells*(2**ndims))]
        f2c = [zefr.iptrToArray(geoAB.f2c, nfaces*2)]

        overFaces = [zefr.iptrToArray(geoAB.overFaces, geoAB.nOverFaces)]
        wallFaces = [zefr.iptrToArray(geoAB.wallFaces, geoAB.nWallFaces)]
        mpiFaces = [zefr.iptrToArray(geoAB.mpiFaces, geoAB.nMpiFaces)]
        procR = [zefr.iptrToArray(geoAB.procR, geo.nMpiFaces)]
        mpiFidR = [zefr.iptrToArray(geoAB.mpiFidR, geo.nMpiFaces)]

        # Wrap relevant solution data
        q = [zefr.dptrToArray(simdata.u_spts, ncells*nspts*nfields)]

        dq = []
        if self.inp.viscous:
          dq.append(zefr.dptrToArray(simdata.dq_spts, ncells*nspts*ndims*nfields))

        self.gridData = {'gridtype' : 'unstructured',
                         'tetConn' : 'None',
                         'pyraConn' : 'None',
                         'prismConn' : 'None',
                         'hexaConn' : c2v,
                         'bodyTag' : [geo.btag],
                         'wallnode' : wallNodes,
                         'obcnode' : overNodes,
                         'grid-coordinates' : xyz,
                         'q-variables' : q,
                         'dq-variables' : dq,
                         'iblanking' : iblank,
                         'scaling' : [self.scale],
                         'face2cell' : f2c,
                         'cell2face' : c2f,
                         'iblank-face' : iblank_face,
                         'iblank-cell' : iblank_cell,
                         'ofaces' : overFaces,
                         'wfaces' : wallFaces,
                         'mfaces' : mpiFaces,
                         'right-proc' : procR,
                         'right-id' : mpiFidR,
                         'nvert-face' : [geoAB.nvert_face],
                         'faceConn' : f2v}
                         #'iblkHasNBHole':0,
                         #'istor':'row'}
                         #'fsicoord':self.fsicoord,
                         #'fsiforce':self.fsiforce,
                         #'fsitag':self.fsitag}  

        if self.inp.motion:
            gridVel = [zefr.dptrToArray(geoAB.grid_vel, nnodes*ndims)]
            offset = [zefr.dptrToArray(geoAB.offset, ndims)]
            Rmat = [zefr.dptrToArray(geoAB.Rmat, ndims*ndims)]

            self.gridData.update({'gridVel',grid_vel,
                'rigidOffset',offset,
                'rigidRotMat',Rmat})

        self.oldCallbacks = {'nodesPerCell',cbs.get_nodes_per_cell,
            'receptorNodes',cbs.get_receptor_nodes,
            'donorInclusionTest',cbs.donor_inclusion_test,
            'donorFrac',cbs.donor_frac,
            'convertToModal',cbs.convert_to_modal}

        self.callbacks = {'nodesPerFace',cbs.get_nodse_per_face,
            'faceNodes',cbs.get_face_nodes,
            'get_q_spt',cbs.get_q_spt,
            'get_q_fpt',cbs.get_q_fpt,
            'get_grad_spt',cbs.get_grad_spt,
            'get_grad_fpt',cbs.get_grad_fpt,
            'q_spts',cbs.get_q_spts,
            'dq_spts',cbs.get_dq_spts}

        if self.useGPU:
            self.callbacks.update({'donorDataDevice',cbs.donor_data_from_device,
                'fringeDataToDevice',cbs.fringe_data_to_device,
                'unblankToDevice',cbs.unblank_data_to_device,
                'faceNodesGPU',cbs.get_face_nodes_gpu,
                'cellNodesGPU',cbs.get_cell_nodes_gpu,
                'q_spts_d',cbs.get_q_spts_d,
                'dq_spts_d',cbs.get_dq_spts_d})

            geoGpu = zefr.get_gpu_geo_data();
            self.gridData.update({'nodesGPU',geoGpu.coord_nodes,
                'eleCoordsGPU',geoGpu.coord_eles,
                'iblankCellGPU',geoGpu.iblank_cell,
                'iblankFaceGPU',geoGpu.iblank_face})

        os.chdir('..')
    
    def writePlotData(self,istep):
        os.chdir('zefr')
        self.z.write_solution()
        os.chdir('..')

    def writeRestartData(self,istep):
        os.chdir('zefr')
        self.z.write_solution()
        os.chdir('..')

    # HELIOS can provide grid velocities
    def setGridSpeeds_maneuver(self,MeshMotionData):
        self.vx = -MeshMotionData['aircraftTranslation'][0]/self.ainf
        self.vy = -MeshMotionData['aircraftTranslation'][1]/self.ainf
        self.vz = -MeshMotionData['aircraftTranslation'][2]/self.ainf
        if MPI.COMM_WORLD.Get_rank() == 0:
            print("gust speed from turns :",self.vx,self.vy,self.vz)

    def deformPart1(self,iter,stage):
        self.z.move_grid(iter,stage)
        pass

    def deformPart2(self,iter,stage):
        self.z.move_grid(iter,stage)
        pass

    def finish(self,step):
        istep = 0
        self.writeRestartData(step)
